File: app/src/main/java/net/joltlang/androidpoc/abiprobe/MainActivity.java
package net.joltlang.androidpoc.abiprobe;

import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.os.Process;
import android.content.SharedPreferences;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

public final class MainActivity extends Activity {
  private JoltRuntime runtime;
  private TextView lifecycle;
  private TextView model;
  private TextView effect;
  private SharedPreferences preferences;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    LinearLayout layout = new LinearLayout(this);
    layout.setOrientation(LinearLayout.VERTICAL);
    lifecycle = new TextView(this);
    model = new TextView(this);
    effect = new TextView(this);
    layout.addView(lifecycle);
    layout.addView(model);
    layout.addView(effect);
    layout.addView(button("Increment", "{:type :counter/inc}"));
    layout.addView(button("Decrement", "{:type :counter/dec}"));
    layout.addView(button("Reset", "{:type :counter/reset}"));
    layout.addView(button("Copy counter", "{:type :platform/copy-counter}"));
    setContentView(layout);

    preferences = getSharedPreferences("jolt-poc", MODE_PRIVATE);
    runtime = new JoltRuntime();
    runtime.dispatch("{:type :lifecycle/create}", ignored -> { });
    if (preferences.contains("counter")) {
      int saved = preferences.getInt("counter", 0);
      runtime.dispatch("{:type :storage/restore :value " + saved + "}", result -> {
        model.setText("Jolt model: " + result);
        effect.setText("Storage restored: " + saved);
      });
    }
  }

  private Button button(String label, String event) {
    Button button = new Button(this);
    button.setText(label);
    button.setOnClickListener(view -> dispatch(event));
    return button;
  }

  private void dispatch(String event) {
    runtime.dispatch(event, result -> {
      model.setText("Jolt model: " + result);
      if (result.contains(":platform/clipboard")) {
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        String text = "Jolt counter: " + counterFrom(result);
        clipboard.setPrimaryClip(ClipData.newPlainText("Jolt counter", text));
        CharSequence copied = clipboard.getPrimaryClip().getItemAt(0).getText();
        effect.setText("Clipboard effect: " + copied);
      } else if (result.contains(":storage/write")) {
        int counter = Integer.parseInt(counterFrom(result));
        preferences.edit().putInt("counter", counter).commit();
        effect.setText("Storage written: " + counter);
      }
    });
  }

  private static String counterFrom(String result) {
    String marker = ":counter ";
    int start = result.indexOf(marker) + marker.length();
    int end = result.indexOf(',', start);
    return result.substring(start, end);
  }

  @Override
  protected void onStart() {
    super.onStart();
    runtime.dispatch("{:type :lifecycle/start}", ignored -> { });
  }

  @Override
  protected void onResume() {
    super.onResume();
    runtime.dispatch("{:type :lifecycle/resume}", result -> {
      lifecycle.setText("Runtime: initialized\n"
          + "Process ABI: " + Build.SUPPORTED_ABIS[0] + "\n"
          + "PID: " + Process.myPid() + "\n"
          + "Jolt thread: HandlerThread(JoltRuntime)\n"
          + "Lifecycle model: " + result);
      model.setText("Jolt model: " + result);
    });
  }

  @Override
  protected void onDestroy() {
    if (runtime != null) runtime.close();
    super.onDestroy();
  }
}

File: app/src/main/java/net/joltlang/androidpoc/abiprobe/MainActivity.kt
package net.joltlang.androidpoc.abiprobe

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.VibrationEffect
import android.os.Vibrator
import android.Manifest
import android.content.pm.PackageManager
import android.app.NotificationChannel
import android.app.NotificationManager
import androidx.core.app.NotificationCompat
import java.util.Locale
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.NavigationBar
import androidx.compose.material3.NavigationBarItem
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp

class MainActivity : ComponentActivity() {
  private lateinit var runtime: JoltRuntime
  private var debugEvalServer: DebugEvalServer? = null
  private var output by mutableStateOf("Starting Jolt runtime…")

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    runtime = JoltRuntime()
    if (BuildConfig.DEBUG) try { debugEvalServer = DebugEvalServer(runtime) } catch (_: Exception) { }
    setContent { DiagnosticApp() }
    dispatch(":lifecycle/create")
  }

  private fun dispatch(type: String) = runtime.dispatch("{:type $type}") {
    output = it
    when {
      it.contains(":platform/vibrate") -> {
        (getSystemService(Vibrator::class.java)).vibrate(VibrationEffect.createOneShot(50, VibrationEffect.DEFAULT_AMPLITUDE))
        output = "$it\nVibration executed"
      }
      it.contains(":platform/open-uri") -> {
        startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://jolt-lang.net")))
        output = "$it\nURL intent launched"
      }
      it.contains(":platform/read-info") -> output = "$it\nLocale: ${Locale.getDefault()}\nPackage: $packageName"
      it.contains(":permission/request") -> requestPermissions(arrayOf(Manifest.permission.POST_NOTIFICATIONS), 7)
      it.contains(":notification/show") -> {
        val manager = getSystemService(NotificationManager::class.java)
        manager.createNotificationChannel(NotificationChannel("jolt", "Jolt", NotificationManager.IMPORTANCE_DEFAULT))
        if (checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED) {
          manager.notify(7, NotificationCompat.Builder(this, "jolt").setSmallIcon(android.R.drawable.ic_dialog_info).setContentTitle("Jolt").setContentText("Counter notification").setAutoCancel(true).build())
          output = "$it\nNotification posted"
        } else output = "$it\nNotification denied"
      }
    }
  }

  @androidx.compose.runtime.Composable
  private fun DiagnosticApp() {
    var screen by remember { mutableStateOf("Runtime") }
    val screens = listOf("Runtime", "State", "Lifecycle", "Effects", "Permission", "Persistence", "Worker")
    Scaffold(bottomBar = { NavigationBar { screens.forEach { name ->
      NavigationBarItem(selected = screen == name, onClick = { screen = name }, icon = {}, label = { Text(name) })
    } } }) { padding ->
      Column(Modifier.fillMaxSize().padding(padding).padding(16.dp)) {
        Text(screen)
        Text(output)
        when (screen) {
          "Runtime" -> { Text(runtime.diagnostics()); Button(onClick = { dispatch(":lifecycle/resume") }) { Text("Refresh runtime") } }
          "State" -> listOf(":counter/inc" to "Increment", ":counter/dec" to "Decrement", ":counter/reset" to "Reset").forEach { (event, label) -> Button(onClick = { dispatch(event) }) { Text(label) } }
          "Lifecycle" -> listOf(":lifecycle/create", ":lifecycle/start", ":lifecycle/resume", ":lifecycle/pause", ":lifecycle/stop").forEach { event -> Button(onClick = { dispatch(event) }) { Text(event) } }
          "Effects" -> {
            Button(onClick = { dispatch(":platform/copy-counter") }) { Text("Copy counter") }
            Button(onClick = { dispatch(":platform/vibrate") }) { Text("Vibrate") }
            Button(onClick = { dispatch(":platform/open-url") }) { Text("Open Jolt URL") }
            Button(onClick = { dispatch(":platform/read-info") }) { Text("Read locale/package") }
            Button(onClick = { dispatch(":platform/notify-counter") }) { Text("Show notification") }
          }
          "Permission" -> Button(onClick = { dispatch(":permission/request-notifications") }) { Text("Request notifications") }
          "Persistence" -> Button(onClick = { dispatch(":storage/restore :value 0") }) { Text("Restore counter") }
          "Worker" -> Button(onClick = { dispatch(":worker/completed") }) { Text("Complete worker") }
        }
      }
    }
  }

  override fun onStart() { super.onStart(); dispatch(":lifecycle/start") }
  override fun onResume() { super.onResume(); dispatch(":lifecycle/resume") }
  override fun onPause() { dispatch(":lifecycle/pause"); super.onPause() }
  override fun onStop() { dispatch(":lifecycle/stop"); super.onStop() }
  override fun onDestroy() { debugEvalServer?.close(); runtime.close(); super.onDestroy() }
}
